hw/top_earlgrey/data/clocks.xdc

```diff
@@ -72,39 +90,166 @@ set_input_delay  -add_delay -clock_fall -clock jtag_tck -min  0.0 [get_ports {IO
 set_input_delay  -add_delay -clock_fall -clock jtag_tck -max 15.0 [get_ports {IOR0 IOR2}]
 
 ## SPI clocks
-create_clock -add -name clk_spi  -period 100.00 -waveform {0 50} [get_ports SPI_DEV_CLK]
-set_input_delay -clock clk_spi 5 [get_ports SPI_DEV_D0]
-set_input_delay -clock clk_spi 5 [get_ports SPI_DEV_D1]
-set_output_delay -clock clk_spi 5 [get_ports SPI_DEV_D0] -add_delay
-set_output_delay -clock clk_spi 5 [get_ports SPI_DEV_D1] -add_delay
-
-# set clock sense on the input to spi buffers to help the tool understand the clocks are shifted versions of each other
-# This can also be accomplished through create_genearted_clocks.
-
-## set clock sense approach
-##set_clock_sense -negative \
-##  [get_pins -filter {DIRECTION == OUT && IS_LEAF} -of_objects \
-##    [get_nets -segments -of_objects \
-##      [get_pins top_*/u_spi_device/gen_fpga_buf.gen_bufr.bufr_i_i_1/I] \
-##    ] \
-##  ] \
-##  -clocks clk_spi
-##
-##set_clock_sense -positive \
-##  [get_pins -filter {DIRECTION == OUT && IS_LEAF} -of_objects \
-##    [get_nets -segments -of_objects \
-##      [get_pins top_*/u_spi_device/gen_fpga_buf.gen_bufr.bufr_i_i_1__0/I] \
-##    ] \
-##  ] \
-##  -clocks clk_spi
-
-## create_generated_clock appraoch
-## create_generated_clock is preferred since the buffer cell used here is hand-instantiated, while the set_clock_sense point is simply a LUT
+set spi_dev_period 100.0
+set spi_dev_half_period [expr ${spi_dev_period} / 2]
+# Max board skew between signals
+set spi_dev_board_skew  0.5
+# Max board delay
+set spi_dev_board_delay 0.6
+# Board skew affects input path for sampling
+set spi_dev_in_delay_min [expr -2.0 - ${spi_dev_board_skew}]
+set spi_dev_in_delay_max [expr  3.0 + ${spi_dev_board_skew}]
+# The board delay affects time remaining on the output path.
+set spi_dev_out_hold      -5.0
+set spi_dev_out_setup     [expr  5.0 + 2 * ${spi_dev_board_delay}]
+
+create_clock -add -name clk_spi  -period ${spi_dev_period} \
+    -waveform "0 ${spi_dev_half_period}" [get_ports SPI_DEV_CLK]
+# CSB must act as a clock, in addition to data and a reset.
+create_clock -name clk_spid_csb -period ${spi_dev_period} \
+    -waveform "${spi_dev_half_period} [expr ${spi_dev_half_period} + 1]" \
```

Comment:
Is this waveform chosen arbitrarily?

Reply:
Yeah, I should've written a comment. This clock definition + the source clock latency constraints below basically say, "CSB is data that sometimes is used to clock things."

The waveform was semi-arbitrarily chosen to show that both edges happen around the (phantom) falling edge of clk_spi / SPI_DEV_CLK, and the source clock latency functions the same way as `set_input_delay`. It does not reflect the duty cycle or minimum pulse width of the real CSB.

We might actually need to adjust this for the ASIC, since this pulse width may or may not be valid there. This worked around limitations in Vivado's timing engine, but the ASIC tool's may be able to handle a different approach.

Reply:
Ok, let's maybe add that in a follow up so that we can land the RTL change.